### PR TITLE
Patch indexmanager

### PR DESCRIPTION
**Pull request by AymericNoel**
### Now Client is functional and some server routes too

What I've done : 
- Add BasePath injection in Trakx.IndexManager.Api because call were never made because of that
- Add Cookie secure policy in order to get authentication with openId
- Update Trakx.Exchange.ApiClient package version because the old one had errors in `AddTrakxExchangeApi()` with Iconfiguration in parameter

### CODE REVIEW

File: src/Trakx.IndexManager.ApiClient/AddIndexManagerApiClientExtension.cs [hunk 1]
```diff
@@ -23,11 +24,13 @@ public static partial class AddIndexManagerApiClientExtension
             var apiConfiguration = configuration.GetSection(nameof(IndexManagerApiConfiguration))
                 .Get<IndexManagerApiConfiguration>();
             Guard.Against.NullOrEmpty(apiConfiguration.BaseUrl, nameof(apiConfiguration.BaseUrl));
-            serviceCollection.AddSingleton<ICredentialsProvider, GoogleCredentialProvider>();
-            
+            var options = Options.Create(apiConfiguration);
```
Comment (monsieurleberre) [inline src/Trakx.IndexManager.ApiClient/AddIndexManagerApiClientExtension.cs:27]: I think this is the less verbose way to do that:
1) make sure the BaseUrl is decorated with a [Required]
Then:
```suggestion
            services.AddOptions();
            services.Configure<IndexManagerApiConfiguration>(
                Configuration.GetSection(nameof(IndexManagerApiConfiguration)));
```

File: src/Trakx.IndexManager.ApiClient/AuthorisedClient.cs [hunk 2]
```diff
@@ -8,12 +8,16 @@ namespace Trakx.IndexManager.ApiClient
     internal abstract class AuthorisedClient
     {
         private readonly ICredentialsProvider _credentialsProvider;
+        public IndexManagerApiConfiguration Configuration { get; protected set; }
+        private string? _baseUrl;
```
Comment (monsieurleberre) [inline src/Trakx.IndexManager.ApiClient/AuthorisedClient.cs:12]: so when is this _baseUrl ever set?

File: src/Trakx.IndexManager.ApiClient/AuthorisedClient.cs [hunk 3]
```diff
@@ -8,12 +8,16 @@ namespace Trakx.IndexManager.ApiClient
     internal abstract class AuthorisedClient
     {
         private readonly ICredentialsProvider _credentialsProvider;
+        public IndexManagerApiConfiguration Configuration { get; protected set; }
```
Comment (monsieurleberre) [inline src/Trakx.IndexManager.ApiClient/AuthorisedClient.cs:11]: Can't `IndexManagerApiConfiguration` that be readonly?

File: src/Trakx.IndexManager.ApiClient/ClientConfigurator.cs [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+using Microsoft.Extensions.Options;
+using Trakx.Utils.Api;
+
+namespace Trakx.IndexManager.ApiClient
+{
+    internal class ClientConfigurator
+    {
+        public ClientConfigurator(IServiceProvider provider)
+        {
+            Configuration = provider.GetService<IOptions<IndexManagerApiConfiguration>>()!.Value;
```
Comment (monsieurleberre) [inline src/Trakx.IndexManager.ApiClient/ClientConfigurator.cs:12]: Just fyi, generally injecting the service provider itself is considered an anti-pattern.
I am aware this is done in the Trakx.Exchange.ApiClient, the reason being that the ApiKeyCredentialProvider relies on the PublicClient to get the time on which to base its signature, so we can't inject it into itself...
I dont know if it is possible here, but the preferred way is to inject an `IOptions<IndexManagerApiConfiguration>` and an `ICredentialProvider` (If the whole project needs only one, this should be OK).
Not a big deal 😉

File: src/Trakx.IndexManager.Client/appsettings.json [hunk 5]
```diff
@@ -8,7 +8,7 @@
   },
   "AllowedHosts": "*",
   "IndexManagerApiConfiguration": {
-    "BaseUrl": "https://localhost:44330",
+    "BaseUrl": "https://localhost:44366",
```
Comment (monsieurleberre) [inline src/Trakx.IndexManager.Client/appsettings.json:11]: This keeps changing if you debug using the IIS server. The other way is to debug the projet itself after configuration the launchsetttings once ad for all.
No big deal either :)

**Review by monsieurleberre — APPROVED** [review]

File: src/Trakx.IndexManager.Client/appsettings.json (hunk 5 above)
Comment (AymericNoel) [reply src/Trakx.IndexManager.Client/appsettings.json:11]: Thanks !

File: src/Trakx.IndexManager.ApiClient/AddIndexManagerApiClientExtension.cs [hunk 6]
```diff
@@ -24,13 +24,14 @@ public static partial class AddIndexManagerApiClientExtension
...
-            serviceCollection.AddSingleton(options);
+            serviceCollection.AddOptions();
+            serviceCollection.Configure<IndexManagerApiConfiguration>(
+                configuration.GetSection(nameof(IndexManagerApiConfiguration)));
 
             serviceCollection.AddClients();
             serviceCollection.AddAuthenticationDependencies(configuration);
 
-            serviceCollection.AddSingleton(s => new ClientConfigurator(s));
+            serviceCollection.AddSingleton(s => new ClientConfigurator(s.GetService<IOptions<IndexManagerApiConfiguration>>()!,s.GetService<ICredentialsProvider>()!));
```
Comment (monsieurleberre) [inline src/Trakx.IndexManager.ApiClient/AddIndexManagerApiClientExtension.cs:34]: sorry to be a pain: don't do this manually
- IOptions<IndexManagerApiConfiguration> is already registered for resolution by the line 27/28
- register the resolution of ICredentialsProvider by providing the type you want
- register the ClientConfigurator
==> after that you have (given the call to AddClients()) all the dependencies necessary for the injection framework to automatically resolve the clients, client configurator, etc, without having to specify things by explicitly calling the constructor of ClientConfigurator

Ask me on slack if it doesn't make sense ;)

File: src/Trakx.IndexManager.ApiClient/AuthorisedClient.cs [hunk 7]
```diff
@@ -8,16 +8,15 @@ namespace Trakx.IndexManager.ApiClient
     internal abstract class AuthorisedClient
     {
         private readonly ICredentialsProvider _credentialsProvider;
-        public IndexManagerApiConfiguration Configuration { get; protected set; }
-        private string? _baseUrl;
-        protected string BaseUrl => _baseUrl ??= Configuration!.BaseUrl;
+        public readonly IndexManagerApiConfiguration Configuration;
```
Comment (monsieurleberre) [inline src/Trakx.IndexManager.ApiClient/AuthorisedClient.cs:11]: clean as a whistle :) 🧹 👍

**Review by monsieurleberre — COMMENTED** [review]
Almost!!!!!

File: src/Trakx.IndexManager.ApiClient/AddIndexManagerApiClientExtension.cs (hunk 6 above)
Comment (AymericNoel) [reply src/Trakx.IndexManager.ApiClient/AddIndexManagerApiClientExtension.cs]: Yeah of course ! It makes sense and it is logical !

File: src/Trakx.IndexManager.ApiClient/AddIndexManagerApiClientExtension.cs [hunk 8]
```diff
@@ -31,7 +30,7 @@ public static partial class AddIndexManagerApiClientExtension
             serviceCollection.AddClients();
             serviceCollection.AddAuthenticationDependencies(configuration);
 
-            serviceCollection.AddSingleton(s => new ClientConfigurator(s.GetService<IOptions<IndexManagerApiConfiguration>>()!,s.GetService<ICredentialsProvider>()!));
+            serviceCollection.AddSingleton<ClientConfigurator>();
```
Comment (monsieurleberre) [inline src/Trakx.IndexManager.ApiClient/AddIndexManagerApiClientExtension.cs:33]: Perfect 👍

**Review by monsieurleberre — APPROVED** [review]

**Pull request merged**